### ISSUE TITLE
DrBBDup: The DynamoRIO Basic Block Duplication Extension

### ISSUE THREAD

**Issue by johnfxgalea** [DynamoRIO/dynamorio#4134]
This is the main issue page concerning the DynamoRIO Basic Block Duplication Extension: a code builder that duplicates code of basic blocks and dispatches control according to runtime conditions so that different instrumentation of the same basic blocks may be efficiently executed.

During the app2app stage, the code of a basic block is duplicated and prepended to the original fragment in order to generate multiple copies. DrBBDup manages these copies by maintaining book-keep data per fragment and inserting dispatching code to execute the appropriate basic block copy.

For example, lets say we have the following bb:

```
mov ebx ecx
mov esi eax
ret
```

At a very high-level, DrBBDup produces code as follows

```
cmp [eax] 0x00000000
jz LABEL 2
  
LABEL 1 // BB Version 1
_INSTRUM CODE A_
mov ebx ecx
_INSTRUM CODE A_
mov esi eax
jmp EXIT LABEL

LABEL 2 // BB Version 2
_INSTRUM CODE B_
mov ebx ecx
_INSTRUM CODE B_
mov esi eax

EXIT LABEL
ret /* Not duplicated to abide by DR bb rules */
```

The number of basic block versions is customisable to suit the user's needs. In the example the dup count is 1 because we produce only one additional copy.

Apart from the main functionality of DrBBDup, this issue also encompasses the following:
1.  Dynamic generation of basic block copies (contingent on #3778 being merged).
2. Gathering of general statistics, such as tracking which cases are executed most frequently.
3. Support for Arm.
4. Include sample program.

There are a number of applications which I built using the extension. For instance, with two bb copies, one can turn on/off instruction tracing via nudges during runtime. One basic block copy is instrumented with tracing code, and another version with no instrumentation at all. Depending on a flag, the user can control which basic block copy is executed at runtime, without the need to undergo expensive flushing.

**Comment by derekbruening** [comment]
This should be useful for #3995.  Currently for a single swap mid-run, a full flush is used, but the request is to support many swaps.  Presumably multi-version instru will be the better approach there.  I was going to implement it inside the tracer, but having a library sounds nicer.

**Comment by johnfxgalea** [comment]
Updated description so that issue also concerns the inclusion of a simple use-case of drbbdup in the sample directory.

**Comment by johnfxgalea** [comment]
Make use of  XINST_CREATE in order to prepare for ARM support.

https://github.com/DynamoRIO/dynamorio/pull/4192#discussion_r394073409

**Comment by derekbruening** [comment]
@johnfxgalea  -- I started looking at using drbbdup in drcachesim for #3995.  I hit an issue early on:

User data param to analysis callback: drbbdup_analyzer_orig_t takes its user_data from drbbdup_init().  But drbbdup_init() can only be called once up front: so this user data does not seem very useful as it's essentially global.  But the more important issue is that drcachesim today creates its user data in the app2app event, stores there whether it did a drutil_expand_rep_string or not, and passes that to the analysis function's user data.  I'm not seeing how to do that with the drbbdup framework.

**Comment by derekbruening** [comment]
See the app2app and analysis functions here: https://github.com/DynamoRIO/dynamorio/blob/master/clients/drcachesim/tracer/tracer.cpp#L1171

**Comment by derekbruening** [comment]
Another issue: drcachesim is using drmgr_is_first_nonlabel_instr() in its insert-instru event.  I guess it can record instrlist_first_nonlabel at the end of each case analysis callback.

**Comment by derekbruening** [comment]
Another issue: drcachesim needs its insert instrumentation callback to go *after* drwrap's (see https://github.com/DynamoRIO/dynamorio/blob/master/clients/drcachesim/tracer/tracer.cpp#L162).  It looks like drbbdup hardcodes the insert event to be at DRMGR_PRIORITY_DRBBDUP = -1500, which is too early since DRMGR_PRIORITY_INSERT_DRWRAP is 500.

I'm not sure I can work around this one.  The app2app I could move the user data into TLS, and the first_nonlabel I could store myself.  But the priority does not seem solvable from the outside.

I guess this raises a larger issue: we want the drwrap clean call to only be inserted for one of our drbbdup cases.  But with no control over drwrap, we can't arrange that, and drwrap is going to go insert its clean call at the start before the drbbdup case dispatch I would guess.

**Comment by derekbruening** [comment]
How to solve?

If drbbdup were integrated inside drmgr: would non-bbdup-aware users of
drmgr like drwrap be invoked for the default case insertion and not the
other cases?  So the user can only have wrapping for just one case (has
to be default)?  Provide an option where user can pick one case, or all
cases?

Or, drwrap has to turn its control model around and have the user call
"do_drwrap_instru" from its insertion event, with drwrap not registering
instru events (but still registering modload, etc.)?

**Comment by johnfxgalea** [comment]
Thanks, I am looking into this. Regarding drwrap - are its clean calls instrumented heavily or can you turn off drbbdup (for now) for those basic blocks which drwrap will instrument via drbbdup' set_up callback? I think you can use drwrap_is_post_wrap(instr_get_app_pc()

**Comment by derekbruening** [comment]
Can't turn it off: if the tracer is in do-not-trace mode we can't have a snippet of trace suddenly being recorded b/c we turned off the dispatcher.

**Comment by johnfxgalea** [comment]
> Another issue: drcachesim is using drmgr_is_first_nonlabel_instr() in its insert-instru event. I guess it can record instrlist_first_nonlabel at the end of each case analysis callback.

drbbdup already provides drbbdup_is_first_instr(). I can add drbbdup_is_first_nonlabel() also that takes into account drbbdup cases.

**Comment by derekbruening** [comment]
> > Another issue: drcachesim is using drmgr_is_first_nonlabel_instr() in its insert-instru event. I guess it can record instrlist_first_nonlabel at the end of each case analysis callback.
> 
> drbbdup already provides drbbdup_is_first_instr(). I can add drbbdup_is_first_nonlabel() also that takes into account drbbdup cases.

So we have several things provided by drmgr (app2app creating user data; this first_nonlabel utility) that we end up having to duplicate for drbbdup.  It would be nice to avoid the duplication and the user having to change what they call and having weird bugs if they call the wrong one (maybe do what drmgr does with macros to detect calling the wrong thing?).  But I'm not sure there is a good solution other than somehow integrating drbbdup into drmgr and making the duplication part of drmgr, which has its own complications.

**Comment by johnfxgalea** [comment]
> But the more important issue is that drcachesim today creates its user data in the app2app event, stores there whether it did a drutil_expand_rep_string or not, and passes that to the analysis function's user data. I'm not seeing how to do that with the drbbdup framework.

What I can do is implement another call back in the app2app stage right before bb duplication, getting user data  and providing it to other callbacks. However, I figured that clients wishing such functionality just can register an app2app event with a higher priority than drbbdup.

Which do you prefer?

**Comment by derekbruening** [comment]
> > But the more important issue is that drcachesim today creates its user data in the app2app event, stores there whether it did a drutil_expand_rep_string or not, and passes that to the analysis function's user data. I'm not seeing how to do that with the drbbdup framework.
> 
> What I can do is implement another call back in the app2app stage right before bb duplication, getting user data and providing it to other callbacks. However, I figured that clients wishing such functionality just can register an app2app event with a higher priority than drbbdup.

I'm not following the last part here: a client can register an app2app, but how do they get the user data created in their app2app to their drbbdup analysis callbacks?

**Comment by johnfxgalea** [comment]
Sorry, I mean via TLS.

**Comment by johnfxgalea** [comment]
>But I'm not sure there is a good solution other than somehow integrating drbbdup into drmgr and making the duplication part of drmgr, which has its own complications.

I did indeed think of this. One complication for instance is drbbdup's reliance on drreg however.

**Comment by derekbruening** [comment]
So the workaround from above:
> The app2app I could move the user data into TLS

Re: the app2app priority: shouldn't drbbdup have a very *low* (i.e., a high number: the opposite of what it has in the code today) priority for app2app?  It should duplicate *after* all transformations like repstr expansion.  Nobody wants an app2app xform happening *after* the bbdup, right?  I would expect it to go last.

I'm not sure what the best thing is: we now have multiple exts that are not just utility functions called by the client but which provide their own callbacks, making it hard to compose them together.  But to turn them into utilities makes for a more difficult usage model for the typical user.  Maybe they could have two modes, one where the user has to call all their instru events in the right sequence, as suggested above for drwrap.

**Comment by johnfxgalea** [comment]
> So the workaround from above:
> 
> > The app2app I could move the user data into TLS
> 
> Re: the app2app priority: shouldn't drbbdup have a very _low_ (i.e., a high number: the opposite of what it has in the code today) priority for app2app? It should duplicate _after_ all transformations like repstr expansion. Nobody wants an app2app xform happening _after_ the bbdup, right? I would expect it to go last.

Yes absolutely, that should be avoided. For my work, I just put other transformations with a higher priority that drbbdups, but happy to lower it.

**Comment by johnfxgalea** [comment]
> So the workaround from above:
> 
> > The app2app I could move the user data into TLS
> 
> Re: the app2app priority: shouldn't drbbdup have a very _low_ (i.e., a high number: the opposite of what it has in the code today) priority for app2app? It should duplicate _after_ all transformations like repstr expansion. Nobody wants an app2app xform happening _after_ the bbdup, right? I would expect it to go last.
> 
> I'm not sure what the best thing is: we now have multiple exts that are not just utility functions called by the client but which provide their own callbacks, making it hard to compose them together. But to turn them into utilities makes for a more difficult usage model for the typical user. Maybe they could have two modes, one where the user has to call all their instru events in the right sequence, as suggested above for drwrap.

Yes, I agree, extensions implementing their own call-backs does make integration currently difficult. If drreg provided utility functions, along with their management routines triggered externally, I think drbbdup could be integrated to drmgr. But like you said, such a usage model places some burden on the user.

**Comment by johnfxgalea** [comment]
> User data param to analysis callback: drbbdup_analyzer_orig_t takes its user_data from drbbdup_init(). But drbbdup_init() can only be called once up front: so this user data does not seem very useful as it's essentially global.

Oh lastly, regarding above. I try to avoid globals to improve modularity - hence the passing of user data.

**Comment by derekbruening** [comment]
> > User data param to analysis callback: drbbdup_analyzer_orig_t takes its user_data from drbbdup_init(). But drbbdup_init() can only be called once up front: so this user data does not seem very useful as it's essentially global.
> 
> Oh lastly, regarding above. I try to avoid globals to improve modularity - hence the passing of user data.

I would expect the user data to come from set_up_bb_dups.  Then it can be local to that bb.  It could still be the same thing passed to the init routine if someone wants, but there's flexibility to make it local.

**Comment by johnfxgalea** [comment]
Yea, so these as separate - data which transcends over all bb fragments and data that is local to a bb. We could try and join the two, but the user would have to create a struct to store their refs.

**Comment by derekbruening** [comment]
The two flags `for_trace` and `translating` are missing.  Many clients don't care (including drcachesim) but some do.  Might be an issue with complex clients in the future.

**Comment by derekbruening** [comment]
Ditto for the dr_emit_flags_t return value.

**Comment by derekbruening** [comment]
I would want to either provide a TLS address for the encoding, or query a TLS offset, to avoid the mem2mem copy of my encoding value to the drbbdup slot at the top of every single bb.

**Comment by johnfxgalea** [comment]
> I would want to either provide a TLS address for the encoding, or query a TLS offset, to avoid the mem2mem copy of my encoding value to the drbbdup slot at the top of every single bb.

I don't think inclusion is a problem and will include. However, I am not entirely understanding your use case. There is drbbdup_get_encoding_opnd(). Can't you set the encoding via an immediately store?

**Comment by derekbruening** [comment]
> > I would want to either provide a TLS address for the encoding, or query a TLS offset, to avoid the mem2mem copy of my encoding value to the drbbdup slot at the top of every single bb.
> 
> I don't think inclusion is a problem and will include. However, I am not entirely understanding your use case. There is drbbdup_get_encoding_opnd(). Can't you set the encoding via an immed store? Or do you want to encode once and only update drbbdup's slot on demand?

Actually yes I can write directly.  Maybe drbbdup can document that it guarantees to not write to the slot itself, so users know they can have insert_encode be a nop (why not let it be NULL?)

**Comment by derekbruening** [comment]
I accidentally explicitly registered the default case and it let me and then called my case callbacks twice with the same encoding value -- sthg to have it check for and return an error I would guess.

**Comment by derekbruening** [comment]
> > Another issue: drcachesim is using drmgr_is_first_nonlabel_instr() in its insert-instru event. I guess it can record instrlist_first_nonlabel at the end of each case analysis callback.
> 
> drbbdup already provides drbbdup_is_first_instr(). I can add drbbdup_is_first_nonlabel() also that takes into account drbbdup cases.

I think this will have to be added: I tried implementing it myself in the user by recording the actual instr pointer in the orig analysis but that does not match anything, so it's not clear how to do it in the client w/o knowledge of how drbbdup is doing copying and what goes where.

**Comment by derekbruening** [comment]
The tracer does some analysis where it inserts labels to store information about adjacent app instrs.  What will happen to these labels?  If it does this analysis in the case cb, will they be there in the insert cb?  I assume so and that the only problem would be if they were added in the orig cb, which is after the instrs are cloned?

**Comment by johnfxgalea** [comment]
Are these labels inserted in the app2app stage? If yes, they should be there.

**Comment by derekbruening** [comment]
> Are these labels inserted in the app2app stage? If yes, they should be there.

No, the analysis stage.  It can be the per-case cb for my use case.

**Comment by johnfxgalea** [comment]
> > > I would want to either provide a TLS address for the encoding, or query a TLS offset, to avoid the mem2mem copy of my encoding value to the drbbdup slot at the top of every single bb.
> > 
> > 
> > I don't think inclusion is a problem and will include. However, I am not entirely understanding your use case. There is drbbdup_get_encoding_opnd(). Can't you set the encoding via an immed store? Or do you want to encode once and only update drbbdup's slot on demand?
> 
> Actually yes I can write directly. Maybe drbbdup can document that it guarantees to not write to the slot itself, so users know they can have insert_encode be a nop (why not let it be NULL?)

So I do not need to include anything right (except for docs)?  There are  drbbdup_get_encoding_opnd() and   drbbdup_set_encoding() which supports your case right?

**Comment by derekbruening** [comment]
> > > > I would want to either provide a TLS address for the encoding, or query a TLS offset, to avoid the mem2mem copy of my encoding value to the drbbdup slot at the top of every single bb.
> > > 
> > > 
> > > I don't think inclusion is a problem and will include. However, I am not entirely understanding your use case. There is drbbdup_get_encoding_opnd(). Can't you set the encoding via an immed store? Or do you want to encode once and only update drbbdup's slot on demand?
> > 
> > 
> > Actually yes I can write directly. Maybe drbbdup can document that it guarantees to not write to the slot itself, so users know they can have insert_encode be a nop (why not let it be NULL?)
> 
> So I do not need to include anything right (except for docs)? There are drbbdup_get_encoding_opnd() and drbbdup_set_encoding() which supports your case right?

Yes.  Maybe also allow insert_encode to be NULL.  Maybe document that the slot is initialized to 0?  Or maybe not and make the client always set it.

**Comment by johnfxgalea** [comment]
> > Are these labels inserted in the app2app stage? If yes, they should be there.
> 
> No, the analysis stage. It can be the per-case cb for my use case.

Ah, no they won't be there. I thought one should avoid insertion during the analysis stage and only store results in TLS/user data.

I'll try and add support for it if it really required.

**Comment by derekbruening** [comment]
> > > Are these labels inserted in the app2app stage? If yes, they should be there.
> > 
> > 
> > No, the analysis stage. It can be the per-case cb for my use case.
> 
> Ah, no they won't be there. I thought one should avoid insertion during the analysis stage and only store results in TLS/user data.

Avoid insertion of regular instructions, yes.  Labels though are often used with note and instr_get_label_data_area() data (https://dynamorio.org/dynamorio_docs/page_drmgr.html#sec_drmgr_notes) for per-instruction info, which is not easy to store separately.

> 
> I'll try and add support for it if it really required.

drcachesim's instrumentation optimizations require labels.  Is there any complication?  Is the instrlist passed to the analysis case cb really not the same as the one actually used for insertion?

**Comment by johnfxgalea** [comment]
> > > > Are these labels inserted in the app2app stage? If yes, they should be there.
> > > 
> > > 
> > > No, the analysis stage. It can be the per-case cb for my use case.
> > 
> > 
> > Ah, no they won't be there. I thought one should avoid insertion during the analysis stage and only store results in TLS/user data.
> 
> Avoid insertion of regular instructions, yes. Labels though are often used with note and instr_get_label_data_area() data (https://dynamorio.org/dynamorio_docs/page_drmgr.html#sec_drmgr_notes) for per-instruction info, which is not easy to store separately.
> 
> > I'll try and add support for it if it really required.
> 
> drcachesim's instrumentation optimizations require labels. Is there any complication? Is the instrlist passed to the analysis case cb really not the same as the one actually used for insertion?

Okay I'll add this to my list then.

Basically, in order to separate concerns, drbbdup creates and passes a new cloned instrlist so that call-back functions may easly consider code of only one basic block rather than the original working instrlist containing duplications.

**Comment by derekbruening** [comment]
Lack of ARM and AArch64 support is also a blocker for fully adopting for drcachesim.  Ideally the existing instru would all be under the same framework, without separate events for x86 vs arm, and with existing delayed-trace features switching from flushing to drbbdup without suddenly only working on x86.  While thinking about arm: the direct encoding store and read will probably need barriers there.

**Comment by johnfxgalea** [comment]
I don't think ARM is particularly hard and will also include on the list. drbbdup does not insert many instructions of its own.

**Comment by derekbruening** [comment]
Summarizing to get an x86-only prototype in drcachesim: need is_first_nonlabel and need case-analysis-cb-added labels to persist.  Then can try to get to endpoint to ensure there are no further issues.

**Comment by derekbruening** [comment]
I see that drbbdup has a fragment deletion event.  We just hit perf problems with that: #4226.  That may delay official adoption into drcachesim, but certainly I will try to get a functional prototype to iron out the interface, as this is an impl issue solved either by speeding up the event or by using a different scheme in drbbdup (drmem and drwrap all abandoned fragment deletion use for invalidation b/c it's racy and instead invalidate on new bbs; I don't remember how drbbdup is using it).

**Comment by johnfxgalea** [comment]
> I see that drbbdup has a fragment deletion event. We just hit perf problems with that: #4226. That may delay official adoption into drcachesim, but certainly I will try to get a functional prototype to iron out the interface, as this is an impl issue solved either by speeding up the event or by using a different scheme in drbbdup (drmem and drwrap all abandoned fragment deletion use for invalidation b/c it's racy and instead invalidate on new bbs; I don't remember how drbbdup is using it).

drbbdup maintains book-keeping data per basic block to keep track which case encodings are registers to the particular basic block. I use fragment deletion to track whether such book-keeping data is valid and delete it if necessary.

**Comment by derekbruening** [comment]
> > I see that drbbdup has a fragment deletion event. We just hit perf problems with that: #4226. That may delay official adoption into drcachesim, but certainly I will try to get a functional prototype to iron out the interface, as this is an impl issue solved either by speeding up the event or by using a different scheme in drbbdup (drmem and drwrap all abandoned fragment deletion use for invalidation b/c it's racy and instead invalidate on new bbs; I don't remember how drbbdup is using it).
> 
> drbbdup maintains book-keeping data per basic block to keep track which case encodings are registered to the particular basic block. I use fragment deletion to track whether such book-keeping data is valid and delete it if necessary.

Yes, that is what it was intended for, and what Dr. Memory and drwrap and other clients started using it for, but as I mentioned they hit problems and for many cases it turns out to be better to invalidate in the block creation events instead.  That is what drwrap does, and what DrM does for some of its state, though some remains in the fragment delete event with a counter.

I suspect that by removing the actual fragment deletion stuff the #4226 perf issue could be alleviated though I have not studied or measured it.

**Comment by johnfxgalea** [comment]
Yep, I think I can also arrange the code to get around fragment delete. I just mentioned how drrbbdup currently uses fragment delete to address the below comment (and not to defend it).

> I don't remember how drbbdup is using it).

**Comment by derekbruening** [comment]
I may get a chance to try a quick #4226 fix today; if that works then we're all set.  If not we can evaluate options then.

**Comment by derekbruening** [comment]
I think the major missing piece for drbbdup is AArchXX support.  The missing pieces are summarized at https://github.com/DynamoRIO/dynamorio/issues/3995#issuecomment-1007024517: basically, loading the encoding and comparing the encoding to the constants.

I would think that once that's done we can probably close this issue and file separate issues on any further enhancements or problems with drbbdup.

**Comment by derekbruening** [comment]
OK I also see that `opts.atomic_load_encoding` is not implemented at all

**Comment by derekbruening** [comment]
@johnfxgalea I see that drbbdup_duplicate_phase() returns `DR_EMIT_STORE_TRANSLATIONS` -- is that due  to dynamic case handling?  If there's no dynamic case handling it wouldn't need to store it?  Is there actual non-determinism?  I see a flush for dynamic case handling: if there's a clean slate after the flush the instrumentation doesn't vary otherwise?

Asking b/c DR and drreg's support for storing translations is not as good as for recreating.

**Comment by johnfxgalea** [comment]
I think  `DR_EMIT_DEFAULT` should work for the simple non-dynamic case and recreation should be deterministic. Let's go ahead with the switch as it saves memory as well, and use `DR_EMIT_STORE_TRANSLATIONS` when dynamic case handling is required.

**Comment by derekbruening** [comment]
I think we can consider this base feature issue to be complete.  Let's file new issues on future changes/improvements.
